dask/layers.py

```diff
@@ -132,13 +132,72 @@ def _dict(self):
         return self._cached_dict
 
     def __getitem__(self, key):
-        return self._dict[key]
+        """Fetch / materialize a single item"""
```

Comment:
Is `__getitem__` used for the final materialization of `ArrayOverlapLayer`, or do we effectively rebuild the entire graph later anyway? Nothing is cached here, so if a `key in layer` check triggers the graph construction below, it will just be forgotten after returning, no? Do we also need an `ArrayOverlapLayer.__contains__` definition to avoid the pre-mature graph construction?

Reply:
`__getitem__` is only used to materialize a single {key:value} pair, which is fast for individual access. When the full layer needs to be materialized, it still uses the old `_construct_graph` function to perform the materialization. 

A better way to do this would be to redefine `_dict` to just call `__getitem__` for all keys in `_dask_keys`, and for `__getitem__` to cache any computed pair in the `_cached_dict` attribute. I'm unsure of exactly the amount that would slow down `_construct_graph` off the top of my head - but since `_construct_graph` is more or less just a for loop over keys anyway, I'm pretty sure it wouldn't be a significant difference. 

To your other question: no, since `__getitem__` only materializes one value at a time rather than the full layer, the `__contains__` method comes for free.

Reply:
However, if we use the redefinition of `_dict` as I said above with a partially `_cached_dict` value, this would then interfere with the meaning of `is_materialized` as it's currently defined. As in, materializing a single key would then lead to `is_materialized` being True...which isn't really true, as it's currently intended. I'm not entirely sure how the materialized flag is used elsewhere in the codebase, so I don't know what the ramifications of that would be.

Reply:
I'm pretty sure that I agree with everything you said, but I'm still a bit unclear on whether it is necessary to materialize a single k-v pair for the purposes of checking if a key is present in the layer.  There is a good chance that I'm missing something, but my understanding is that you only *use* this new logic for the `key in layer` check, because the actual graph-materialization will always pass through `_construct_graph`.

Is there any way to define `__contains__` in a way that would be faster than materializing the subgraph needed to produce a given key? Maybe not?